### PR TITLE
[4.x] Add option to append original filename to Glide URLs

### PR DESCRIPTION
**Pull request by ryanmitchell**
This PR adds a new config option `statamic.assets.image_manipulation.append_original_filename` which when set to `true` adds the original filename to the generated Glide URLs.

This works when both caching and not-caching URLs.

Take this tag:

`{{ glide src="/assets/content/lagoa.jpg" width="200" dpr="2" }}`

With caching disabled:
Before this PR: `/img/asset/YXNzZXRzL2NvbnRlbnQvbGFnb2EuanBn?w=200&dpr=2&s=8d3a50041d9e25a78e287bddb2b7a2d1`
After this PR: `/img/asset/YXNzZXRzL2NvbnRlbnQvbGFnb2EuanBn/lagoa.jpg?w=200&dpr=2&s=8d3a50041d9e25a78e287bddb2b7a2d1`

With caching enabled:
Before this PR: `/img/containers/assets/content/lagoa.jpg/62e0a4215262774b8a7c175f457d9be1.jpg`
After this PR: `/img/containers/assets/content/lagoa.jpg/62e0a4215262774b8a7c175f457d9be1-lagoa.jpg`

Fixes: https://github.com/statamic/cms/issues/2814

### CODE REVIEW

**Comment by robdekort** [comment]
It feels like Christmas this week. Thanks Ryan. This is wonderful and seems to work like charm. I wonder if it needs to be an option though. I can't think of any reason why you wouldn't want the filename appended by default.

However, I suspect the whole image cache would have to be rebuild with this on. Which could be a reason for it to be opt-in only on existing sites. True by default on new sites?

**Comment by jonassiewertsen** [comment]
That's looking wonderful.

Is this better for SEO? Just wondering and no SEO export:
After this PR: `/img/containers/assets/content/lagoa.jpg/lagoa-62e0a4215262774b8a7c175f457d9be1.jpg`

**Comment by jasonvarga** [comment]
Beautiful Ryan. Looks like the `setCachePathCallable` was only added to Glide this year. I was wondering why we didn't do this from the start.

@robdekort Yeah I think this could be the default for new sites - i.e. set it to true in statamic/statamic, and then make it the actual default in statamic/cms for v5.

@jonassiewertsen Are you suggesting just to put the filename before the hash instead of after? Or am I missing something else?

**Comment by ryanmitchell** [comment]
Anyone we can ask? - it’s an easy change but would be good to get it right from the start.

**Comment by jonassiewertsen** [comment]
Jup, I think so too. That's why I am so annoying asking that question without knowing the answer :-D

**Comment by robdekort** [comment]
I’m suspecting prepending is indeed better. Google always weighs first words heavier. Good one!

Also: sounds like nice path forward @jasonvarga

**Comment by jasonvarga** [comment]
> and then make it the actual default in statamic/cms for v5.

Or maybe even just make it do it all the time. Not an option.

**Comment by sjclark** [comment]
https://developers.google.com/search/docs/appearance/google-images#descriptive-titles-captions-filenames
>...the filename can give Google very light clues about the subject matter of the image. When possible, use filenames that are short, but descriptive. For example, my-new-black-kitten.jpg is better than IMG00023.JPG. Avoid using generic filenames like image1.jpg, pic.gif, 1.jpg when possible.

Alt tags and figcaptions are **far** more important/weighted (some Google staff reference this on various podcasts), but still a great idea 😄 For me personally I think the above is also likely to fix various file encoding errors. Lots of libraries don't correctly parse files when they don't have a file extension appended (like glightbox)

File: src/Imaging/GlideUrlBuilder.php [hunk 1]
```diff
@@ -64,4 +74,13 @@ public function build($item, $params, $filename = null)
 
         return URL::prependSiteRoot($builder->getUrl($path, $params));
     }
+
+    private function optionallySetFilename($filename)
```
Comment (edalzell) [inline src/Imaging/GlideUrlBuilder.php:78]: typehints?

Comment (ryanmitchell) [reply src/Imaging/GlideUrlBuilder.php]: Good spot, thanks

**Comment by drpark** [comment]
As i'm working with seo experts, they always told me to do
this : /img/containers/assets/content/lagoa.jpg/62e0a4215262774b8a7c175f457d9be1/lagoa.jpg

instead of this:
With caching enabled:
After this PR: /img/containers/assets/content/lagoa.jpg/62e0a4215262774b8a7c175f457d9be1-lagoa.jpg

so would be great to be able to replace the last - by a / even with a config param

**Comment by ryanmitchell** [comment]
At the moment it returns /img/containers/assets/content/lagoa.jpg/lagoa-62e0a4215262774b8a7c175f457d9be1.jpg which from what I've discussed with our SEO guys is the right approach.

**Comment by jasonvarga** [comment]
I can see where excluding the hash from the filename itself would be nice from a purely aesthetic standpoint.

e.g. If you right click the image and downloaded it, it would save as `lagoa.jpg` instead of `lagoa-62e0a4215262774b8a7c175f457d9be1.jpg`.

Would having the hash as a directory instead of the filename affect the functionality?

**Comment by ryanmitchell** [comment]
No I don't think it would - should be a very small change to implement.

**Review by duncanmcclean — CHANGES_REQUESTED** [review]

**Review by duncanmcclean — APPROVED** [review]

**Comment by jackmcdade** [comment]
Thanks for all the work on this! I've added making this the default (and only option) for v5.

**Pull request merged**